controllers/class.tapestry-controller.php

```diff
@@ -0,0 +1,337 @@
+<?php
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-errors.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-helpers.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-user-roles.php";
+require_once dirname(__FILE__) . "/../utilities/class.tapestry-node-permissions.php";
+require_once dirname(__FILE__) . "/../interfaces/interface.tapestry-controller.php";
+
+/**
+ * Add/update/retrieve a Tapestry
+ * 
+ */
+class TapestryController implements ITapestryController
+{
+    private $postId;
+
+    /**
+     * Constructor
+     * 
+     * @param   Number  $postId post ID
+     * 
+     * @return  NULL
+     */
+    public function __construct($postId = 0)
+    {
+        $this->postId = (int) $postId;
+    }
+
+    /**
+     * Add A Tapestry
+     * 
+     * @param   Object  $tapestry   Tapestry
+     * 
+     * @return  Object  $tapestry
+     */
+    public function save($tapestry)
+    {
+        if (empty($tapestry->settings)) {
+            return TapestryErrors::throwsError('SETTINGS_MISSING_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->nodes)) {
+            return TapestryErrors::throwsError('NODES_EXIST_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->groups)) {
+            return TapestryErrors::throwsError('GROUPS_EXIST_IN_NEW_TAPESTRY');
+        }
+        if (!empty($tapestry->links)) {
+            return TapestryErrors::throwsError('LINKS_EXIST_IN_NEW_TAPESTRY');
+        }
+
+        return $this->_addTapestry($tapestry);
+    }
+
+    /**
+     * Retrieve a Tapestry post
+     * 
+     * @return  Object  $tapestry
+     */
+    public function get()
+    {
+        if (!$this->postId) {
+            return TapestryErrors::throwsError('INVALID_POST_ID');
+        }
+
+        return $this->_getTapestry();
+    }
+
+    /**
+     * Add a new node
+     * 
+     * @param   Object  $node   Tapestry node
+     * 
+     * @return  Object  $node   Tapestry node
+     */
+    public function addNode($node)
+    {
+        $tapestryNodeController = new TapestryNodeController($this->postId);
+        $node = $tapestryNodeController->save($node);
+
+        if (is_wp_error($node)) {
+            return $node;
+        }
+
+        $tapestry = get_post_meta($this->postId, 'tapestry', true);
+
+        if (!isset($tapestry->nodes)) {
+            $tapestry->nodes = [];
+        }
+
```

Comment:
Lines 83-87 should not be needed if you initialize and load the tapestry in the constructor.